core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -130,37 +144,167 @@ KOKKOS_INLINE_FUNCTION
+KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_same<T, bool>::value, T>
+    cast_from_bhalf(bhalf_t);
+template <class T>
+KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_same<T, double>::value, T>
+    cast_from_bhalf(bhalf_t);
+template <class T>
+KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_same<T, short>::value, T>
+    cast_from_bhalf(bhalf_t);
+template <class T>
+KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_same<T, int>::value, T>
+    cast_from_bhalf(bhalf_t);
+template <class T>
+KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_same<T, long>::value, T>
+    cast_from_bhalf(bhalf_t);
+template <class T>
+KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_same<T, long long>::value, T>
+    cast_from_bhalf(bhalf_t);
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    std::enable_if_t<std::is_same<T, unsigned short>::value, T>
+        cast_from_bhalf(bhalf_t);
+template <class T>
+KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_same<T, unsigned int>::value, T>
+    cast_from_bhalf(bhalf_t);
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    std::enable_if_t<std::is_same<T, unsigned long>::value, T>
+        cast_from_bhalf(bhalf_t);
+template <class T>
+KOKKOS_INLINE_FUNCTION
+    std::enable_if_t<std::is_same<T, unsigned long long>::value, T>
+        cast_from_bhalf(bhalf_t);
+#endif  // CUDA_VERSION < 11000
+/*********************** END half forward declarations  ***********************/
+
+namespace Impl {
+template <class FloatType>
+class alignas(FloatType) floating_point_wrapper {
  public:
-  using impl_type = Kokkos::Impl::half_impl_t::type;
+  using impl_type = FloatType;
 
  private:
   impl_type val;
+  using fixed_width_integer_type = std::conditional_t<
+      sizeof(impl_type) == 2, uint16_t,
+      std::conditional_t<
+          sizeof(impl_type) == 4, uint32_t,
+          std::conditional_t<sizeof(impl_type) == 8, uint64_t, void>>>;
+  static_assert(!std::is_void<fixed_width_integer_type>::value,
+                "Invalid impl_type");
+
+  // BEGIN: Casting wrappers for supporting multiple impl types
+  static KOKKOS_INLINE_FUNCTION Kokkos::Impl::half_impl_t::type float2impl(
+      float x, Kokkos::Impl::half_impl_t::type&) {
+    return __float2half(x);
+  }
+
+#if CUDA_VERSION >= 11000
+  static KOKKOS_INLINE_FUNCTION Kokkos::Impl::bhalf_impl_t::type float2impl(
+      float x, Kokkos::Impl::bhalf_impl_t::type&) {
+    return __float2bfloat16(x);
+  }
+#endif  // CUDA_VERSION >= 11000
+
+  static KOKKOS_INLINE_FUNCTION float impl2float(
+      Kokkos::Impl::half_impl_t::type x) {
+    return __half2float(x);
+  }
+#if CUDA_VERSION >= 11000
+  static KOKKOS_INLINE_FUNCTION float impl2float(
+      Kokkos::Impl::bhalf_impl_t::type x) {
+    return __bfloat162float(x);
+  }
+#endif  // CUDA_VERSION >= 11000
+
+  template <class T>
+  static KOKKOS_INLINE_FUNCTION Kokkos::Experimental::half_t cast_to_wrapper(
+      T x, volatile Kokkos::Impl::half_impl_t::type&) {
+    return Kokkos::Experimental::cast_to_half(x);
```

Comment:
@nliber: With the latest commit that moved `floating_point_wrapper` into `Kokkos::Experimental::Impl` the compiler can no longer seem to locate the forward decls of `cast_to_half`:
```
kokkos/core/src/Cuda/Kokkos_Cuda_Half.hpp(261): error: no instance of overloaded function "Kokkos::Experimental::cast_to_half" matches the argument list
            argument types are: (Kokkos::Experimental::Impl::floating_point_wrapper<Kokkos::Impl::half_impl_t::type>)
```
Do you know how to provide these forward decls across namespaces?

Reply:
@crtrott: Please let me know if 8516f6e works for you until the forward declaration issue is resolved.